### DISCUSSION TITLE
Better type annotation for `CORSMiddleware`?

### DISCUSSION THREAD

**Discussion by roy-work** [encode/starlette#2722] [Potential Issue] [answered]
See: https://github.com/encode/starlette/blob/master/starlette/middleware/cors.py#L19C1-L19C50

(I'll be discussing `allow_origins`, but the logic herein should apply to `allow_methods`, too.)

`CORSMiddleware` defines the `allow_origins` parameter as,

```python
        allow_origins: typing.Sequence[str] = (),
        allow_methods: typing.Sequence[str] = ("GET",),
```

But the only thing we ever do with `allow_origins` is `… in self.allow_origins`, to test if an origin is allowed. This type annotation, then, _could_ be just `typing.Container[str]`; `Sequence` is an ordered collection, i.e., it requires indexing. This means `Sequence` permits pretty much only `list` & `tuple`, but it would make a lot of sense (and be faster¹) to pass something like a `frozenset` here. (But a `frozenset` is not a `Sequence`.)

<sub>¹`frozenset` `__contains__` checks are about 5× faster on my app's current origin list. But the absolute difference isn't huge: it's 62ns for a list, vs. 11ns for a set, which isn't much in the grand scope of processing an HTTP request. It is more that a set makes more _semantic_ sense to be passable here.</sub>

**Comment by Kludex** [comment, marked as answer]
Okay. PR welcome.
